EGM-certified electron IDs for 2016 data: cut-based and MVA

ikrav opened the pull request:
This PR contains EGM certified cut-based and MVA electron IDs tuned on 80X samples. They are intended to be used on 2016 data. The electron MVA contains the HZZ MVA used for ICHEP16 and the presently recommended general purpose MVA tuned for better performance given features of a more complete 2016 dataset.

This PR is a new version of the recent PR #16839 (about to be closed) that is rebased to CMSSW_9_0_X_2016-12-07-2300. It also contains a few minor changes addressing  comments from reviewers.

ikrav committed:
New HLT-safe electron cut-based selection in VID, incl. python and a few new cut classes

ikrav committed:
Bug fixes, added debug statements

ikrav committed:
A try to fix segv

ikrav committed:
Fixed AOD vs miniAOD detection

ikrav committed:
Corrected cut settings

ikrav committed:
New cut-based electron ID tuned on 80X

ikrav committed:
Fixed a text comment

ikrav committed:
Fixed a typo in cut values

ikrav committed:
Added 80X electron MVA implementation. Restructured location of MVA class files (including old ones) to make things work better for FWLite.

Werbellin committed:
Added general purpose ID

Werbellin committed:
Remove preliminary HZZ setup

Werbellin committed:
Update documentation, bugfixes

ikrav committed:
Adjused MD5 sums of the electron MVA IDs

ikrav committed:
A few changes to make new electron MVA work in FWLite

ikrav committed:
Removed a header from an MVA base class that appears unnecessary and was preventing the release from building in 81X+

ikrav committed:
Merged egm_id_80X_v2_rebased_CMSSW_9_0_X_2016-12-07-2300 from repository ikrav

ikrav committed:
Removed TROOT.h include, appears unnecessary

ikrav committed:
Replaced several import* with explicitly named imports

cmsbuild commented:
A new Pull Request was created by  @ikrav  for CMSSW_9_0_X.

It involves the following packages:

RecoEgamma/EgammaTools
RecoEgamma/ElectronIdentification


@cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@Sam-Harper, @rafaellopesdesa, @lgray this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

cvuosalo commented on RecoEgamma/ElectronIdentification/python/Identification/cutBasedElectronHLTPreselecition_Summer16_V1_cff.py:
```diff
@@ -0,0 +1,99 @@
+from PhysicsTools.SelectorUtils.centralIDRegistry import central_id_registry
+
+# Common functions and classes for ID definition are imported here:
+from RecoEgamma.ElectronIdentification.Identification.cutBasedElectronID_tools import *
```
Still one `import *` in the new code -- please fix like you did the other one. Thanks.

ikrav committed:
Replaced import * with explicit list in a python configuration file.

slava77 commented:
On 12/12/16 12:33 PM, Carl Vuosalo wrote:
> *@cvuosalo* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In
> RecoEgamma/ElectronIdentification/python/Identification/cutBasedElectronHLTPreselecition_Summer16_V1_cff.py
> <https://github.com/cms-sw/cmssw/pull/16979#pullrequestreview-12543328>:
> 
>> @@ -0,0 +1,99 @@
> +from PhysicsTools.SelectorUtils.centralIDRegistry import central_id_registry
> +
> +# Common functions and classes for ID definition are imported here:
> +from RecoEgamma.ElectronIdentification.Identification.cutBasedElectronID_tools import *
> 
> Still one |import *| in the new code -- please fix like you did the
> other one. Thanks.


This import *
cleanup is needed primarily for imports of files which define actual
modules.

It is not needed and is rather tedious to maintain for includes of
functions tools and ingredient PSets.


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16979#pullrequestreview-12543328>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbv6aDVibBdP3Q09j6wci4osQxX0sks5rHa-3gaJpZM4LK6e2>.
>

ikrav commented:
Well, anyways I fixed it already, the list of explicit imports this time is 
not too long. The change is committed. 

> On Dec 12, 2016, at 3:17 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 12/12/16 12:33 PM, Carl Vuosalo wrote:
> > *@cvuosalo* commented on this pull request.
> > 
> > ------------------------------------------------------------------------
> > 
> > In
> > RecoEgamma/ElectronIdentification/python/Identification/cutBasedElectronHLTPreselecition_Summer16_V1_cff.py
> > <https://github.com/cms-sw/cmssw/pull/16979#pullrequestreview-12543328>:
> > 
> >> @@ -0,0 +1,99 @@
> > +from PhysicsTools.SelectorUtils.centralIDRegistry import central_id_registry
> > +
> > +# Common functions and classes for ID definition are imported here:
> > +from RecoEgamma.ElectronIdentification.Identification.cutBasedElectronID_tools import *
> > 
> > Still one |import *| in the new code -- please fix like you did the
> > other one. Thanks.
> 
> 
> This import *
> cleanup is needed primarily for imports of files which define actual
> modules.
> 
> It is not needed and is rather tedious to maintain for includes of
> functions tools and ingredient PSets.
> 
> 
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/cms-sw/cmssw/pull/16979#pullrequestreview-12543328>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AEdcbv6aDVibBdP3Q09j6wci4osQxX0sks5rHa-3gaJpZM4LK6e2>.
> > 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9bc7e83f2a8a590cc2ebfd3b5646036c215fd64d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16979/16976/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16979/16976/summary.html

cvuosalo commented on RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16HZZ.h:
```diff
@@ -0,0 +1,157 @@
+  // MVA name. This is a unique name for this MVA implementation.
+  // It will be used as part of ValueMap names.
+  // For simplicity, keep it set to the class name.
+  const std::string _name = "ElectronMVAEstimatorRun2Spring16HZZ";
```
More prohibited names with leading underbars, lines 132-152.

cvuosalo commented on RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleCalPFClusterIsoCut.cc:
```diff
@@ -0,0 +1,147 @@
+
+private:
+  // Cut values
+  const float _isoCutEBLowPt,_isoCutEBHighPt,_isoCutEELowPt,_isoCutEEHighPt;
```
Again, leading underbars lines 26-31.

cvuosalo commented on RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16GeneralPurpose.h:
```diff
@@ -0,0 +1,154 @@
+  // MVA tag. This is an additional string variable to distinguish
+  // instances of the estimator of this class configured with different
+  // weight files.
+  const std::string _tag;
```
There are still prohibited variables names that start with "_" here and on lines 136-149.

ikrav committed:
Adjusted variable names to comply with CMS coding guidelines that require that data memners do not start from a leading underscore.

cmsbuild commented:
Pull request #16979 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

ikrav commented:
Carl, so I made a number of changes in the new files, including the ones you
pointed out plus a few more, I think. The changes are pushed to GitHub.

However, let me mention: the present set of files new to CMSSW use files from 
older IDs as a starter. There is a ton of those old files in RecoEgamma. Most written by
Lindsey, actually. They are full of leading underscores. I have added some IDs in the past 
in the same style, and  nobody complained until now… I changed only the new files, but 
have not touched any of the old files.

> On Dec 13, 2016, at 4:53 PM, Carl Vuosalo <notifications@github.com> wrote:
> 
> @cvuosalo commented on this pull request.
> 
> In RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16GeneralPurpose.h:
> 
> > +  // Call this function once after the constructor to declare
> +  // the needed event content pieces to the framework
> +  void setConsumes(edm::ConsumesCollector&&) const override final;
> +  // Call this function once per event to retrieve all needed
> +  // event content pices
> +  
> + private:
> +
> +  // MVA name. This is a unique name for this MVA implementation.
> +  // It will be used as part of ValueMap names.
> +  // For simplicity, keep it set to the class name.
> +  const std::string _name = "ElectronMVAEstimatorRun2Spring16GeneralPurpose";
> +  // MVA tag. This is an additional string variable to distinguish
> +  // instances of the estimator of this class configured with different
> +  // weight files.
> +  const std::string _tag;
> 
> There are still prohibited variables names that start with "_" here and on lines 136-149.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 0df48ee21e1b3c4e484e459f1fa4eb14a0d3f5eb

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16979/17035/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>    vars = std::move( packMVAVariables(allMVAVars.see,
           ^
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-14-1100/src/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16GeneralPurpose.cc:386:12: note: remove std::move call here
    vars = std::move( packMVAVariables(allMVAVars.see,
           ^~~~~~~~~~~
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-14-1100/src/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16HZZ.cc:54:28: error: field 'conversionsLabelMiniAOD_' is uninitialized when used here [-Werror,-Wuninitialized]
  conversionsLabelMiniAOD_(conversionsLabelMiniAOD_) {
                           ^
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-14-1100/src/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16HZZ.cc:82:35: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
  const std::vector<float> vars = std::move( fillMVAVariables( particle, iEvent ) );  
                                  ^
</pre>

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16979/17035/summary.html

ikrav commented:
The std::move was there in all previous electron and photon MVA IDs, are we required 
to do anything regarding it this time?

The error of the initialization: I am asking about it the people who wrote this part.
This whole constructor is new in comparison to previous ID versions, so I am not
sure what they intended here. My local build was successful before I submitted the PR,
but it wasn’t a clang build.

> On Dec 14, 2016, at 11:59 AM, cmsbuild <notifications@github.com> wrote:
> 
> -1
> 
> Tested at: 0df48ee
> 
> You can see the results of the tests here:
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16979/17035/summary.html
> 
> I found follow errors while testing this PR
> 
> Failed tests: ClangBuild
> 
> 	• Clang:
> I found a compilation error while trying to compile with clang:
> I used this command:
> scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'
> 
>     vars = std::move( packMVAVariables(allMVAVars.see,
>            ^
> /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-14-1100/src/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16GeneralPurpose.cc:386:12: note: remove std::move call here
>     vars = std::move( packMVAVariables(allMVAVars.see,
>            ^~~~~~~~~~~
> /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-14-1100/src/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16HZZ.cc:54:28: error: field 'conversionsLabelMiniAOD_' is uninitialized when used here [-Werror,-Wuninitialized]
>   conversionsLabelMiniAOD_(conversionsLabelMiniAOD_) {
>                            ^
> /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-14-1100/src/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16HZZ.cc:82:35: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
>   const std::vector vars = std::move( fillMVAVariables( particle, iEvent ) );  
>                                   ^
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ikrav committed:
Fixed a typo introduced with the previous commit when lots of variables got renamed

cmsbuild commented:
Pull request #16979 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

ikrav commented:
The error below is fixed, the change is committed. The cause was my typo when
changing very many variable names to comply CMS coding rules in the previous commit.

> On Dec 14, 2016, at 11:59 AM, cmsbuild <notifications@github.com> wrote:
> 
> /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-14-1100/src/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16HZZ.cc
> :54:28: error: field 'conversionsLabelMiniAOD_' is uninitialized when used here [-Werror,-Wuninitialized]
>   conversionsLabelMiniAOD_(conversionsLabelMiniAOD_) {
> 
>

cvuosalo commented on RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleNormalizedGsfChi2Cut.cc:
```diff
@@ -0,0 +1,46 @@
+  }
+
+private:
+  const double _normalizedGsfChi2CutValueEB,_normalizedGsfChi2CutValueEE,_barrelCutOff;
```
Three more with prohibited leading underbars. Please fix them.

ikrav committed:
Fixed variable names in GsfEleNormalizedGsfChi2Cut to comply CMS coding rules

cmsbuild commented:
Pull request #16979 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

ikrav commented:
This is fixed and committed.

> On Dec 15, 2016, at 3:09 PM, Carl Vuosalo <notifications@github.com> wrote:
> 
> @cvuosalo commented on this pull request.
> 
> In RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleNormalizedGsfChi2Cut.cc:
> 
> > +    CutApplicatorBase(c),
> +    _normalizedGsfChi2CutValueEB(c.getParameter<double>("normalizedGsfChi2CutValueEB")),
> +    _normalizedGsfChi2CutValueEE(c.getParameter<double>("normalizedGsfChi2CutValueEE")),
> +    _barrelCutOff(c.getParameter<double>("barrelCutOff")){    
> +  }
> +  
> +  result_type operator()(const reco::GsfElectronPtr&) const override final;
> +
> +  double value(const reco::CandidatePtr& cand) const override final;
> +
> +  CandidateType candidateType() const override final { 
> +    return ELECTRON; 
> +  }
> +
> +private:
> +  const double _normalizedGsfChi2CutValueEB,_normalizedGsfChi2CutValueEE,_barrelCutOff;
> 
> Three more with prohibited leading underbars. Please fix them.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 8424971c2c661ab4651bfd98eb833cb1f5f97129
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16979/17079/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16979/17079/summary.html

cvuosalo commented:
+1

For #16979 8424971c2c661ab4651bfd98eb833cb1f5f97129

For EGM, adding new certified electron IDs.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_9_0_X_2016-12-15-2300 show no significant differences. An extended test for the previous version of this PR (https://github.com/cms-sw/cmssw/pull/16839#issuecomment-265842966), which differed only slightly from the current one, showed no problems, with no differences and only a very marginal increase in CPU time.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

ikrav commented:
Does this means I may proceed with submitting pull requests to 80X and 81X
with identical content? - thanks

> On Dec 16, 2016, at 11:38 AM, cmsbuild <notifications@github.com> wrote:
> 
> This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
this PR is not merged yet.
You can make backports, but there is a small chance you'd need to edit here and in the backports later.

ikrav commented:
Ok, I take it that I shouldn’t submit new PRs until this PR is merged.
(I sent the email because the previous message said it is fully signed off on).
Higgs people asked me not to do backport to 80X, so instead I will cherry-pick
the commits you asked me to introduce in this PR, and add them to my original
80X branch.


> On Dec 19, 2016, at 1:54 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> this PR is not merged yet.
> You can make backports, but there is a small chance you'd need to edit here and in the backports later.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

davidlange6 commented:
+1
